File: tests/tmpdirs/test-tmpdirs.pl
#!/usr/bin/env perl
# Copyright (C) 2012 Red Hat Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

# Test logic for setting location of tmpdir and cachedir.

use strict;
use warnings;

use Sys::Guestfs;
use File::Temp qw(tempdir);

# Remove any environment variables that may have been set by the
# user or the ./run script which could affect this test.
delete $ENV{LIBGUESTFS_TMPDIR};
delete $ENV{LIBGUESTFS_CACHEDIR};
delete $ENV{TMPDIR};

my $g;

# Defaults with no environment variables set.
$g = Sys::Guestfs->new ();
die unless $g->get_tmpdir () eq "/tmp";
die unless $g->get_cachedir () eq "/var/tmp";

# Create some test directories.
my $a = tempdir (CLEANUP => 1);
my $b = tempdir (CLEANUP => 1);
my $c = tempdir (CLEANUP => 1);

# Setting environment variables.
$ENV{LIBGUESTFS_TMPDIR} = $a;
$ENV{LIBGUESTFS_CACHEDIR} = $b;
$ENV{TMPDIR} = $c;

$g = Sys::Guestfs->new ();
die unless $g->get_tmpdir () eq $a;
die unless $g->get_cachedir () eq $b;

# Creating a handle which isn't affected by environment variables.
$g = Sys::Guestfs->new (environment => 0);
die unless $g->get_tmpdir () eq "/tmp";
die unless $g->get_cachedir () eq "/var/tmp";

# Uses TMPDIR if the others are not set.
delete $ENV{LIBGUESTFS_TMPDIR};
$g = Sys::Guestfs->new ();
die unless $g->get_tmpdir () eq $c;
die unless $g->get_cachedir () eq $b;

delete $ENV{LIBGUESTFS_CACHEDIR};
$g = Sys::Guestfs->new ();
die unless $g->get_tmpdir () eq $c;
die unless $g->get_cachedir () eq $c;

# Directories should be made absolute automatically.
delete $ENV{LIBGUESTFS_TMPDIR};
delete $ENV{LIBGUESTFS_CACHEDIR};
delete $ENV{TMPDIR};
$ENV{TMPDIR} = ".";
$g = Sys::Guestfs->new ();
my $pwd = `pwd`; chomp $pwd;
die unless $g->get_tmpdir () eq $pwd;
die unless $g->get_cachedir () eq $pwd;

File: tests/tmpdirs/test-tmpdirs.py
#!/usr/bin/env python3
# Copyright (C) 2025 Red Hat Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

# Test logic for setting location of tmpdir and cachedir.

import sys
import os
import guestfs
import tempfile

# Remove any environment variables that may have been set by the
# user or the ./run script which could affect this test.
os.environ.pop('LIBGUESTFS_TMPDIR', None)
os.environ.pop('LIBGUESTFS_CACHEDIR', None)
os.environ.pop('TMPDIR', None)

# Defaults with no environment variables set.
g = guestfs.GuestFS()
assert g.get_tmpdir() == "/tmp"
assert g.get_cachedir() == "/var/tmp"

# Create some test directories.
with tempfile.TemporaryDirectory() as a, \
     tempfile.TemporaryDirectory() as b, \
     tempfile.TemporaryDirectory() as c:

    # Setting environment variables.
    os.environ['LIBGUESTFS_TMPDIR'] = a
    os.environ['LIBGUESTFS_CACHEDIR'] = b
    os.environ['TMPDIR'] = c

    g = guestfs.GuestFS()
    assert g.get_tmpdir() == a
    assert g.get_cachedir() == b

    # Creating a handle which isn't affected by environment variables.
    g = guestfs.GuestFS(environment=False)
    assert g.get_tmpdir() == "/tmp"
    assert g.get_cachedir() == "/var/tmp"

    # Uses TMPDIR if the others are not set.
    os.environ.pop('LIBGUESTFS_TMPDIR', None)
    g = guestfs.GuestFS()
    assert g.get_tmpdir() == c
    assert g.get_cachedir() == b

    os.environ.pop('LIBGUESTFS_CACHEDIR', None)
    g = guestfs.GuestFS()
    assert g.get_tmpdir() == c
    assert g.get_cachedir() == c

# Directories should be made absolute automatically.
os.environ.pop('LIBGUESTFS_TMPDIR', None)
os.environ.pop('LIBGUESTFS_CACHEDIR', None)
os.environ.pop('TMPDIR', None)
os.environ['TMPDIR'] = "."
g = guestfs.GuestFS()
pwd = os.getcwd()
assert g.get_tmpdir() == pwd
assert g.get_cachedir() == pwd
